File: py-project/Solution34 2.py
# -*- coding:utf-8 -*-
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # 返回二维列表，内部每个列表表示找到的路径
    def __init__(self):
        self.result = []
        # self.list1 = []

    def FindPathIn(self, root, path, currentNumber,expectNumber):
        currentNumber += root.val
        path.append(root)
        isleaf = root.left == None and root.right == None
        if currentNumber == expectNumber and isleaf:
            onePath = []
            for node in path:
                onePath.append(node.val)
            self.result.append(onePath)

        if currentNumber < expectNumber:
            if root.left:
                self.FindPathIn(root.left, path, currentNumber, expectNumber)
            if root.right:
                self.FindPathIn(root.right, path, currentNumber, expectNumber)
        path.pop()


    def FindPath(self, root, expectNumber):
        # write code here
        if not root:
            return []

        self.FindPathIn(root, [], 0,expectNumber)

        return self.result

if __name__ == '__main__':
    t1 = TreeNode(10)
    t2 = TreeNode(5)
    t3 = TreeNode(12)
    t4 = TreeNode(4)
    t5 = TreeNode(7)

    t1.left = t2
    t1.right = t3
    t2.left = t4
    t2.right = t5

    solution1 = Solution()
    a = solution1.FindPath(t1,22)
    print(a)

File: c2-project/Solution34 2.cpp
//
// Created by 陈高杰 on 2020/2/6.
//
#include <iostream>
#include <stdlib.h>
#include <string>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};

class Solution {
public:
    vector<vector<int> > res;
    void FindPathIn(TreeNode* root,vector<int> path, int expectNumber,int currentNumber){
        currentNumber += root->val;
        path.push_back(root->val);
        bool isleaf = root->left == nullptr && root->right == nullptr;
        if(isleaf && currentNumber ==expectNumber){
            res.push_back(path);
        }
        if(root->left != nullptr){
            FindPathIn(root->left,path,expectNumber,currentNumber);
        }
        if(root->right != nullptr){
            FindPathIn(root->right,path,expectNumber,currentNumber);
        }
        path.pop_back();
        return;
    }
    vector<vector<int> > FindPath(TreeNode* root,int expectNumber) {
        if(root == nullptr){
            return res;
        }
        vector<int> path;
        int  currentNumber = 0;
        FindPathIn(root,path,expectNumber,currentNumber);
        return res;
    }
};

//int main()
//{
//    TreeNode t1 = TreeNode(10);
//    TreeNode t2 = TreeNode(5);
//    TreeNode t3 = TreeNode(12);
//    TreeNode t4 = TreeNode(4);
//    TreeNode t5 = TreeNode(7);
//    t1.left = &t2;
//    t1.right = &t3;
//    t2.left = &t4;
//    t2.right = &t5;
////    cout << t1.right->val << endl;
//    Solution solution;
//    vector<vector<int> > output;
//    output = solution.FindPath(&t1,22);
////    cout << "output:" << output<< endl;
//    system("PAUSE");
//    return 0;
//}
